crates/relayer/src/event/source/rpc.rs

```diff
@@ -0,0 +1,395 @@
+        };
+
+        Ok((source, TxEventSourceCmd(tx_cmd)))
+    }
+
+    pub fn run(mut self) {
+        let _span = error_span!("event_source.rpc", chain.id = %self.chain_id).entered();
+
+        debug!("collecting events");
+
+        let rt = self.rt.clone();
+
+        rt.block_on(async {
+            let mut backoff = poll_backoff(self.poll_interval);
+
+            // Initialize the latest fetched height
+            if let Ok(latest_height) = latest_height(&self.rpc_client).await {
+                self.last_fetched_height = latest_height;
+            }
+
+            // Continuously run the event loop, so that when it aborts
+            // because of WebSocket client restart, we pick up the work again.
+            loop {
+                let before_step = Instant::now();
+
+                match self.step().await {
+                    Ok(Next::Abort) => break,
+
+                    Ok(Next::Continue) => {
+                        // Reset the backoff
+                        backoff = poll_backoff(self.poll_interval);
+
+                        // Check if we need to wait some more before the next iteration.
+                        let delay = self.poll_interval.checked_sub(before_step.elapsed());
+
+                        if let Some(delay_remaining) = delay {
+                            sleep(delay_remaining).await;
+                        }
+
+                        continue;
+                    }
+
+                    Err(e) => {
+                        error!("event source encountered an error: {e}");
+
+                        // Let's backoff the little bit to give the chain some time to recover.
+                        let delay = backoff.next().expect("backoff is an infinite iterator");
+
+                        error!("retrying in {delay:?}...");
+                        sleep(delay).await;
+                    }
+                }
+            }
+        });
+
+        debug!("shutting down event source");
+    }
+
+    async fn step(&mut self) -> Result<Next> {
+        // Process any shutdown or subscription commands before we start doing any work
+        if let Some(next) = self.try_process_cmd() {
+            return Ok(next);
+        }
+
+        let latest_height = latest_height(&self.rpc_client).await?;
+
+        let batches = if latest_height > self.last_fetched_height {
+            trace!(
+                "latest height ({latest_height}) > latest fetched height ({})",
+                self.last_fetched_height
+            );
+
+            self.fetch_batches(latest_height).await.map(Some)?
+        } else {
+            trace!(
+                "latest height ({latest_height}) <= latest fetched height ({})",
+                self.last_fetched_height
+            );
+
+            None
+        };
+
+        // Before handling the batch, check if there are any pending shutdown or subscribe commands.
+        if let Some(next) = self.try_process_cmd() {
+            return Ok(next);
```

Comment:
Why do we need to do this again? Maybe clarify in the comment.

Reply:
ebee52c7